python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +63,1058 @@
+                        # if the formatter read fails we will not be
+                        # caching this file.
+                        cache_msg = " and likely cached"
+
+                    msg = f"(via download to local file{cache_msg})"
+
+                log.debug(
+                    "Reading%s from location %s %s with formatter %s",
+                    f" component {log_component}" if log_component else "",
+                    uri,
+                    msg,
+                    self.name(),
+                )
+
+                with time_this(
+                    log,
+                    msg="Reading%s from location %s %s with formatter %s",
+                    args=(
+                        f" component {log_component}" if log_component else "",
+                        uri,
+                        msg,
+                        self.name(),
+                    ),
+                ):
+                    result = self.read_from_local_file(local_uri, component=component)
+
+                # File was read successfully so can move to cache
+                if can_be_cached:
+                    cache_manager.move_to_cache(local_uri, cache_ref)
+
+        return result
+
+    def read_from_uri(self, uri: ResourcePath, component: str | None = None) -> Any:
+        """Read a dataset from a URI that can be local or remote.
+
+        Parameters
+        ----------
+        uri : `lsst.resources.ResourcePath`
+            URI to use to read the dataset. This URI can be local or remote
+            and can refer to the actual resource or to a locally cached file.
+        component : `str` or `None`, optional
+            The component to be read from the dataset.
+
+        Returns
+        -------
+        in_memory_dataset : `typing.Any`
+            The Python object read from the resource.
+
+        Raises
+        ------
+        FormatterNotImplementedError
+            Raised if there is no support for direct reads from a, possibly,
+            remote URI.
+
+        Notes
+        -----
+        This method is only called if the class property
+        ``can_read_from_uri`` is set to `True`.
+
+        It is possible that a cached local file will be given to this method
+        even if it was originally a remote URI. This can happen if the write
+        resulted in the file being added to the local cache.
+
+        If the full file is being read this file will not be added to the
+        local cache. Consider raising `FormatterNotImplementedError` in
+        this situation, for example if there are no parameters or component
+        specified, and allowing the system to fall back to calling
+        `read_from_local_file` (which will populate the cache if configured
+        to do so).
+        """
+        raise FormatterNotImplementedError("This formatter does not know how to read a file.")
+
+    def read_from_stream(
+        self, stream: BinaryIO | ResourceHandleProtocol, component: str | None = None
```

Comment:
Is it worth passing in the expected size to this method (and read_from_uri) in case they want to decide that it would be more efficient to read from a local file and raise in here? (which would need to be caught above).

Reply:
Added to `read_from_uri` as well. Not added to `read_from_local_file` on the basis that there isn't really much choice even if it's a large file and also if they really need to know the actual size they can ask easily enough and performance shouldn't be an issue. Maybe I should add it for symmetry though.

Reply:
I also added it to `read_from_local_file`...